### PR TITLE
Made OneTime bindings update on DataContext changes

### LINKED ISSUE #17474: Binding on a UserControl inside a DataTemplate broken in 11.2.0
### Describe the bug

`DataContext={Binding}` on a `UserControl` inside a `DataTemplate` is broken in 11.2.0

11.2.0
image

11.1.4
image

### To Reproduce

MainWindow.axaml
```axaml
...
    <ContentControl Content="{Binding Book}">
        <ContentControl.DataTemplates>
            <DataTemplate DataType="m:Book">
                <StackPanel>
                    <TextBlock Background="Red" Text="{Binding}" />

                    <!-- <v:UserControl1 Background="Yellow" /> will be OK -->
                    <!--  however the following will be invisible  -->
                    <v:UserControl1 Background="Yellow" DataContext="{Binding}" />

                    <ContentControl Background="Green" Content="{Binding}">
                        <ContentControl.DataTemplates>
                            <DataTemplate x:DataType="m:Book">
                                <StackPanel>
                                    <TextBlock Text="{Binding}" />
                                </StackPanel>
                            </DataTemplate>
                        </ContentControl.DataTemplates>
                    </ContentControl>
                </StackPanel>
            </DataTemplate>
        </ContentControl.DataTemplates>
    </ContentControl>
...
```
UserControl1.axaml
```axaml
...
    <ContentControl Content="{Binding}">
        <ContentControl.DataTemplates>
            <DataTemplate x:DataType="m:Book">
                <StackPanel>
                    <TextBlock Text="{Binding}" />
                </StackPanel>
            </DataTemplate>
        </ContentControl.DataTemplates>
    </ContentControl>
...
```
MainWindowViewModel.cs
```cs
public partial class MainWindowViewModel : ViewModelBase
{
    public Book Book { get; } = new();
}
```
Book.cs
```cs
public record Book(string Name = "Avalonia", int Price = 4000);
```

### Expected behavior

Binding works properly as in 11.1.4

### Avalonia version

11.2.0

### OS

Windows

### Additional context

May be related to #16242

### LINKED ISSUE #17587: `IsVisible` binding without a path is not applied
### Describe the bug

A binding without a property path applied to `IsVisible` will not be evaluated. The control will remain visible when it should be hidden.

Adding a path of `$self.DataContext` works around the issue.

### To Reproduce

Start Sandbox with this window XAML:

``` xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        x:Class="Sandbox.MainWindow" x:CompileBindings="False">
  <StackPanel>
    <Border IsVisible="{Binding, Converter={x:Static ObjectConverters.IsNotNull}}" Height="19" Background="Red" />
    <Border IsVisible="{Binding $self.DataContext, Converter={x:Static ObjectConverters.IsNotNull}}" Height="19" Background="Green" />
  </StackPanel>
</Window>
```

You will see this:

image

### Expected behavior

The `Border` with the red background should not be visible.

### Avalonia version

11.2.1

### OS

Windows

### Additional context

Not reproducible in 11.1.x

### LINKED ISSUE #17649: View's DataContext is getting set to the parent's view DataContext at initialization, triggering the error "Unable to cast object of type"
### Describe the bug

I've updated my project from Avalonia 11.0.9 to 11.2.2 and my visual tree got broken.
When i got to check out what's actually happening, in the logs the following line appeared:
```
[Binding]An error occurred binding 'ItemsSource' to 'Satellites' at 'Satellites': 'Unable to cast object of type 'Avalonia.Yuujin.SatTrack.ViewModels.MainWindowViewModel' to type 'Avalonia.Yuujin.SatTrack.ViewModels.HomeViewModel'.' (ItemsControl #21626090)
[Binding]An error occurred binding '__AvaloniaReserved::Classes::Hidden' to '!ShowDetails' at 'ShowDetails': 'Unable to cast object of type 'Avalonia.Yuujin.SatTrack.ViewModels.MainWindowViewModel' to type 'Avalonia.Yuujin.SatTrack.ViewModels.HomeViewModel'.' (Border #60417086)
```
And many-many more of similar messages in nested controls.
Resulting in the actual controls' properties not being binded to anything.

'Strange' -- i thought, because in AXAML i am explicitly stating that i want the 'HomeView' to have bind it's `DataContext` to the property `MainWindowViewModel.HomeViewModel`

MainView.axaml
```xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			 xmlns:themes="clr-namespace:Material.Styles.Themes;assembly=Material.Styles"
			 xmlns:views="clr-namespace:Avalonia.Yuujin.SatTrack.Views"
             xmlns:vm="clr-namespace:Avalonia.Yuujin.SatTrack.ViewModels"
             xmlns:md="clr-namespace:Avalonia.Yuujin.SatTrack.ViewModels"
             mc:Ignorable="d" d:DesignWidth="392.73" d:DesignHeight="791.64"
             x:Class="Avalonia.Yuujin.SatTrack.Views.MainView"
             x:DataType="vm:MainWindowViewModel">
	<Design.DataContext>
		<vm:MainWindowViewModel />
	</Design.DataContext>
	
	<Grid Name="MainContainer" RowDefinitions="1*, Auto">
		<TabControl TabStripPlacement="Bottom" SelectedIndex="{Binding SelectedTabIndex}">
			<TabItem>
				<TabItem.Header>
					<StackPanel Orientation="Vertical" Margin="0, 5, 0, 3">
						<Image Height="24" Source="/Assets/home.png"/>
						<TextBlock Text="Home"/>
					</StackPanel>
				</TabItem.Header>
				<views:HomeView DataContext="{Binding HomeViewModel}"/>
			</TabItem>
...
...
```

I've looked up other issues ([#17592](https://github.com/AvaloniaUI/Avalonia/issues/17592))
And thought that similarly, it may have been my own mistake, but no. In the `HomeView.axaml` there isn't a single trace referencing `MainWindowViewModel`

*HomeView.axaml*
```xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			 xmlns:themes="clr-namespace:Material.Styles.Themes;assembly=Material.Styles"
             xmlns:c="clr-namespace:Avalonia.Yuujin.SatTrack.Controls"
             xmlns:vm="clr-namespace:Avalonia.Yuujin.SatTrack.ViewModels"
             xmlns:m="clr-namespace:Avalonia.Yuujin.SatTrack.Models"
             xmlns:cv="clr-namespace:Avalonia.Yuujin.SatTrack.Converters"
             mc:Ignorable="d" d:DesignWidth="392.73" d:DesignHeight="791.64"
             x:Class="Avalonia.Yuujin.SatTrack.Views.HomeView"
			 x:DataType="vm:HomeViewModel">

	<Design.DataContext>
		<vm:HomeViewModel>
			<vm:HomeViewModel.SelectedSatellite>
				<m:Satellite 
					Altitude="3.455735"
					Latitude="66.66666"
					Longitude="-148.2231"/>
			</vm:HomeViewModel.SelectedSatellite>
			<vm:HomeViewModel.MapScale>0.15</vm:HomeViewModel.MapScale>
		</vm:HomeViewModel>
	</Design.DataContext>

	<UserControl.Resources>
		<cv:DegreeToMinutesRepresentationConverter x:Key="DegreesConverter"/>
	</UserControl.Resources>

	<Grid>
		<ScrollViewer VerticalScrollBarVisibility="Hidden">
			<ItemsControl ItemsSource="{Binding Satellites}">
				<ItemsControl.ItemTemplate>
					<DataTemplate>
						<Border Margin="0,5" BoxShadow="0 5 5 0 #181818">
							<c:SatelliteSummary Height="64" DataContext="{Binding}"/>
						</Border>
					</DataTemplate>
				</ItemsControl.ItemTemplate>
			</ItemsControl>
		</ScrollViewer>

		<Border Background="{StaticResource MaterialPaperBrush}" 
				Classes.Hidden="{Binding !ShowDetails}" 
				IsVisible="{Binding ShowDetails}"
				PointerPressed="DetailsPointerPressed">
			<Border.Styles>
				<Style Selector="Border">
					<Setter Property="RenderTransform" Value="translate(0px, 0px)"/>
				</Style>
				<Style Selector="Border.Hidden">
					<Setter Property="RenderTransform" Value="translate(393px, 0px)"/>
				</Style>
			</Border.Styles>
			<Border.Transitions>
				<Transitions>
					<TransformOperationsTransition Property="RenderTransform" Duration="0:0:0.5" Easing="CubicEaseOut"/>
				</Transitions>
			</Border.Transitions>

			<Grid RowDefinitions="48, Auto, 1*" ClipToBounds="True" ShowGridLines="False">
				<Grid.Styles>
					<Style Selector="TextBlock">
						<Setter Property="FontSize" Value="20"/>
						<Setter Property="FontFamily" Value="{StaticResource SourceCodePro}"/>
					</Style>
				</Grid.Styles>
				
				<Button HorizontalAlignment="Left" VerticalAlignment="Stretch" Command="{Binding HideDetails}" Width="48" Classes="Flat" CornerRadius="0">
					<Image Source="/Assets/arrow-left.png" Width="24"/>
				</Button>
				<Button HorizontalAlignment="Right" VerticalAlignment="Stretch" Command="{Binding ShowDetailsSettings}" Width="48" Classes="Flat" CornerRadius="0">
					<Image Source="/Assets/format-list-bulleted.png" Width="24"/>
				</Button>
				<TextBlock Text="Details: NOAA-19" VerticalAlignment="Center" Margin="48,0,0,2" FontSize="22"/>

				<StackPanel Orientation="Vertical" Grid.Row="1">
					<TextBlock Text="{Binding SelectedSatellite.Altitude, StringFormat='Altitude:  {0:000.000} km'}" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="{Binding SelectedSatellite.Latitude, 
							   StringFormat='Latitude:  {0}', 
							   Converter={StaticResource DegreesConverter}
							   ConverterParameter=Latitude}" VerticalAlignment="Center" Margin="36,0,0,2"/>
					
					<TextBlock Text="{Binding SelectedSatellite.Longitude, 
							   StringFormat='Longitude: {0}', 
							   Converter={StaticResource DegreesConverter}
							   ConverterParameter=Longitude}" VerticalAlignment="Center" Margin="36,0,0,2"/>
					
					<TextBlock Text="--------------------------" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="{Binding SelectedSatellite.OrbitalSpeed, StringFormat='Oribtal speed: {0:0.00000} km/s'}" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="Radial speed:  7.55300 km/s" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="--------------------------" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="Azimuth:   150.23° (SSE)" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="Elevation:  20.23°" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="--------------------------" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="Next pass:     01h 22m 00s" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="Max Elevation: 88.12°" VerticalAlignment="Center" Margin="36,0,0,2"/>
					<TextBlock Text="--------------------------" VerticalAlignment="Center" Margin="36,0,0,2"/>
				</StackPanel>

				<Border Background="#1E1E1E" Grid.Row="2">
					<Canvas x:Name="MapCanvas" Background="#003243" ClipToBounds="True" 
							PointerWheelChanged="Canvas_PointerWheelChanged"
							PointerPressed="Canvas_PointerPressed_1"
							PointerReleased="Canvas_PointerReleased"
							PointerMoved="Canvas_PointerMoved"
							Gestures.Pinch="Canvas_Pinch"
							Gestures.PinchEnded="Canvas_PinchEnd">
						<Canvas.GestureRecognizers>
							<PinchGestureRecognizer/>
						</Canvas.GestureRecognizers>

						<Image 
							x:Name="MapImage"
							Source="/Assets/miller2.png" 
							Canvas.Top="{Binding ImgPointPosY}" 
							Canvas.Left="{Binding ImgPointPosX}" 
							Width="{Binding CanvasMapImageWidth}"/>
						
						
						<ItemsControl
							ItemsSource="{Binding SatellitePathPoints}"
							Canvas.Top="{Binding ImgPointPosY}"
							Canvas.Left="{Binding ImgPointPosX}" >

							<ItemsControl.ItemsPanel>
								<ItemsPanelTemplate>
									<Grid/>
								</ItemsPanelTemplate>
							</ItemsControl.ItemsPanel>

							<ItemsControl.ItemTemplate>
								<DataTemplate>
									<Polyline Points="{Binding}" Stroke="#F06548" StrokeThickness="3"/>
								</DataTemplate>
							</ItemsControl.ItemTemplate>
						</ItemsControl>
						
						<Ellipse Fill="Red" 
								 Width="10" Height="10"
								 RenderTransform="translate(-5px, -5px)"
								 Canvas.Top="{Binding SatPointPosY}" Canvas.Left="{Binding SatPointPosX}"/>
						
						<Polygon Points="{Binding LitAreaPoints}" Stroke="DarkBlue" StrokeThickness="1" 
								 Fill="#60F06548" 
								 Canvas.Top="{Binding ImgPointPosY}"
								 Canvas.Left="{Binding ImgPointPosX}"/>
					</Canvas>
				</Border>

				<Border 
					Grid.Row="0" 
					Grid.RowSpan="2" 
					Background="#1E1E1E"
					HorizontalAlignment="Right"
					VerticalAlignment="Top"
					Classes.Hidden="{Binding !ShowDetailsMenu}"
					IsVisible="{Binding ShowDetailsMenu}">
					<Border.Styles>
						<Style Selector="Border">
							<Setter Property="RenderTransform" Value="translate(0px, 0px)"/>
						</Style>
						<Style Selector="Border.Hidden">
							<Setter Property="RenderTransform" Value="translate(300px, 0px)"/>
						</Style>
					</Border.Styles>
					<Border.Transitions>
						<Transitions>
							<TransformOperationsTransition Property="RenderTransform" Duration="0:0:0.5" Easing="CubicEaseOut"/>
						</Transitions>
					</Border.Transitions>

					<StackPanel>
						<StackPanel.Styles>
							<Style Selector="Button">
								<Setter Property="Background" Value="Transparent"/>
								<Setter Property="CornerRadius" Value="0"/>
							</Style>
							<Style Selector="Button /template/ ContentPresenter#PART_ContentPresenter">
								<Setter Property="HorizontalContentAlignment" Value="Left"/>
							</Style>
							<Style Selector="TextBlock">
								<Setter Property="Foreground" Value="#FFFFFF"/>
								<Setter Property="FontSize" Value="18"/>
								<Setter Property="TextAlignment" Value="Left"/>
								<Setter Property="Margin" Value="12,6"/>
							</Style>
						</StackPanel.Styles>

						<Grid>
							<TextBlock>Update period: 01s</TextBlock>
							<Button Classes="Flat"/>
						</Grid>
						<Grid>
							<TextBlock>SGP4 period: 60s</TextBlock>
							<Button Classes="Flat"/>
						</Grid>
						<Grid>
							<TextBlock>Update SGP4</TextBlock>
							<Button Classes="Flat"/>
						</Grid>
						<Grid>
							<TextBlock>Update TLE</TextBlock>
							<Button Classes="Flat"/>
						</Grid>
						<Grid>
							<TextBlock>Remove Record</TextBlock>
							<Button Classes="Flat" Command="{Binding RemoveSatRecord}"/>
						</Grid>
					</StackPanel>
				</Border>
				
			</Grid>
		</Border>
	</Grid>
</UserControl>
```

And the actual property also has the appropriate type:
```CSharp
private HomeViewModel homeViewModel;
public HomeViewModel HomeViewModel { get => homeViewModel; set => this.RaiseAndSetIfChanged(ref homeViewModel, value); }
```

So i went do debug, subscribed to event `DataContextChanged` in the HomeView.axaml.cs class like so:
```CSharp
public HomeView()
{
    InitializeComponent();

    this.DataContextChanged += HomeView_DataContextChanged;
}

private void HomeView_DataContextChanged(object? sender, System.EventArgs e)
{
    if (DataContext is HomeViewModel vm)
    {
        ;
    }
}
```

in the debugger i found that the HomeView's `DataConext` is actually being set to the `MainWindowViewModel`
image
*what? how? why?*, -- i went and check the stack trace, to find out where that has happened.

I found that it's happening in the setter of `MainWindow.DataConext` in the `OnFrameworkInitializationComplete` method in `App.cs` class
image
I've additionally made a sanity check that `MainWindow` is not, in fact, a `HomeView` 

After that, the `DataContext` is being set properly. But all of the errors are already happened.
image

And so, as the result, the `DataContext` of the controls is being set properly at the end of the day, but the bindings on the controls properties are broken, and are not resolved automatically on `DataContext` change

HomeView's `DataContext` is set properly
image

Custom control's `DataContext` is set properly
image

The data is present
image

But the `Text` property on `TextBlock` inside a custom control is Unset
image

### To Reproduce

1. Create a parent View or a Window
2. Create a child view with controls
3. Bind the child view's control data to the data in the child's viewmodel
4. Place the child view in the parent view or a window and explicitly set the child's DataContext to property declared in the parent's viewmodel
5. Set the parent view/window data context to new object in `OnFrameworkInitializationCompleted` in `App.cs`

### Expected behavior

The children controls `DataContext` not being set to the parent `DataContext`, but immediately to the appropriate property, if it is present.
--or--
The binding being restored when the `DataContext` changes to the object with appropriate type. 

### Avalonia version

11.2.2

### OS

Windows, Android

### Additional context

SatteliteSummary.axaml
```xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:m="clr-namespace:Avalonia.Yuujin.SatTrack.Models"
             mc:Ignorable="d" d:DesignWidth="392.7" d:DesignHeight="64"
			 x:DataType="m:Satellite"
             x:Class="Avalonia.Yuujin.SatTrack.Controls.SatelliteSummary"
			 ClipToBounds="True">

	<Design.DataContext>
		<m:Satellite>
			<m:Satellite.SatcatRecord>
				<m:SatcatRecord 
					ObjectName="SUOMI UWP [+]"
					NORADCatId="12345"/>
			</m:Satellite.SatcatRecord>
		</m:Satellite>
	</Design.DataContext>
	
	<Grid 
		ShowGridLines="False"
		ClipToBounds="True"
		ColumnDefinitions="6, 48, 1*, 48, 6">
		
		<Image Grid.Column="1"
			   Height="32"
			   Width="32"
			   Source="/Assets/satellite-variant.png"/>

		<Grid Grid.Column="2"
			  ShowGridLines="False"
			  ClipToBounds="True"
			  RowDefinitions="Auto, 1*"
			  Margin="0,5">
			<Grid.Styles>
				<Style Selector="TextBlock">
					<Setter Property="FontSize" Value="15"/>
					<Setter Property="FontFamily" Value="{StaticResource SourceCodePro}"/>
				</Style>
			</Grid.Styles>
			
			<TextBlock Text="{Binding SatcatRecord.ObjectName}" 
					   HorizontalAlignment="Left" 
					   VerticalAlignment="Center" 
					   Margin="9,0,0,0"/>
			<TextBlock Text="{Binding SatcatRecord.NORADCatId, StringFormat='NORAD: {0}'}" 
					   HorizontalAlignment="Right" 
					   VerticalAlignment="Center" 
					   Background="{StaticResource MaterialPaperBrush}"
					   Margin="0,0,3,0"/>

			<TextBlock Text="{Binding QuickStatus}"
					   Grid.Row="1"
					   FontSize="20"
					   Foreground="{Binding QuickStatusColor}"
					   HorizontalAlignment="Right"
					   VerticalAlignment="Top"
					   Margin="0,-4,3,0"/>
			<TextBlock Text="{Binding QuickElevation}"
					   Grid.Row="1"
					   FontSize="12"
					   Foreground="{Binding QuickStatusColor}"
					   HorizontalAlignment="Right"
					   VerticalAlignment="Bottom"
					   Margin="0,0,3,1"/>

		</Grid>

		<Button
			Grid.Column="3"
			CornerRadius="100"
			Height="40"
			Width="40"
			Classes.Flat="True"
			Margin="2,0,0,0"
			Command="{Binding Details}">
			<Image
				Height="50"
				Width="70"
				Source="/Assets/menu-right.png"/>
		</Button>
	</Grid>
</UserControl>
```

### PR DESCRIPTION
**Pull request by MrJul**
## What does the pull request do?
This PR changes how bindings with a `OneTime` mode work, to behave the same way as in WPF (and as currently documented), updating their target not only a single time, but once per data context change.

In addition, `null` data contexts are now considered valid values for bindings without a path.

## What is the current behavior?
`OneTime` bindings are evaluated only once.
Bindings without a path (`{Binding}`) are considered "pending" (they don't update their target nor call their converter) while their data context is `null`.

## What is the updated/expected behavior with this PR?
`OneTime` bindings are evaluated once _per data context_, like in WPF.
`null` is now a valid `DataContext` for pathless bindings, and will use the configured `TargetNullValue` and `Converter`, also matching WPF.

## Notes
As explained in https://github.com/AvaloniaUI/Avalonia/issues/17587#issuecomment-2493943363, the two changes are entangled: with the previous behavior we can't have `null` be a valid data context without breaking `OneTime` bindings using them.

Unit tests have been added.
One unit test has been deleted, since one-time bindings now don't stop on the first update.

## Fixed issues
 - Fixes #17587
 - Fixes #17474
 - Fixes #17649

### CODE REVIEW

File: src/Avalonia.Base/Data/Core/BindingExpression.cs [hunk 1]
```diff
@@ -231,10 +233,14 @@ internal void OnNodeValueChanged(int nodeIndex, object? value, Exception? dataVa
...
         if (nodeIndex == _nodes.Count - 1)
         {
-            // If the binding source is a data context without any path and is currently null, treat it as an invalid
-            // value. This allows bindings to DataContext and DataContext.Property to share the same behavior.
-            if (value is null && _nodes[nodeIndex] is DataContextNodeBase)
-                value = AvaloniaProperty.UnsetValue;
+            if (_mode == BindingMode.OneTime)
+            {
+                // In OneTime mode, only changing the data context updates the binding.
+                if (!_shouldUpdateOneTimeBindingTarget && _nodes[nodeIndex] is not DataContextNodeBase)
```
Comment (maxkatz6) [inline src/Avalonia.Base/Data/Core/BindingExpression.cs:239]: Is this data context only, or also any source changes?
For example, if I set relative source to TemplatedParent, and then move control to another template.
Or set DynamicResource as a source and change it, without any path.

**Comment by timunie** [comment]
This probably closes more issues.

File: src/Avalonia.Base/Data/Core/BindingExpression.cs (hunk 1 above)
Comment (MrJul) [reply src/Avalonia.Base/Data/Core/BindingExpression.cs:239]: I've just checked WPF: `RelativeSource` bindings never update, _even in OneWay_ mode. I've tried `AncestorType` and `TemplatedParent` then moving controls from one container to another. And a `DynamicResource` can only be applied to a `DependencyProperty`.

So we should choose a behavior for Avalonia here. It's a pretty niche use case, I'd say we keep the documented behavior of _once per DataContext_, not _once per source_.

**Comment by MrJul** [comment]
> This probably closes more issues.
> 
> #17474 #17649

Thanks! I've tested the two repros and they now work properly with the changes from this PR.
I've added them to the OP.

**Comment by MrJul** [comment]
Added as a backport candidate due to the fix for `DataContext` bindings, an issue that regularly pops up.

**Review by grokys — APPROVED** [review]

**Pull request merged**